### PR TITLE
Refacto to pypi package 

### LINKED ISSUE #202: Move functions out of main.py
The `main.py` have a lot of functions that could be moved away such as `run_scenario `wich could be static method for the `scenario.py` file.

### PR DESCRIPTION
**Pull request by arthurPignet**
This PR follows the [PR205](https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/205).

Branch 'Moving Function' rebased from master. 

Original PR text : 

Moving functional block to more appropriate files in order to provide a coherent library.

Change the folder architecture to clean the crowded root folder.

fix #202

### CODE REVIEW

File: tests/unit_tests.py
Comment (bowni) [inline tests/unit_tests.py:176]: Sure about this indentation?

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:8]: Ok - maybe add anchor links here

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:20]: Adding a link would be perfect here

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:90]: Ok, that works. At some point in the future we might want to extract the parameters descriptions from the main `README.md` and make it a dedicated document or page

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:147]: A link would be good here

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:149]:
```suggestion
## Contributivity measurement methods
```

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:151]:
```suggestion
To use contributivity measurement tools, you will have to set parameters to your `scenario` object.
```

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:162]: Same here - a link would be great

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:174]:
```suggestion
*Work in progress*
```

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:2]:
```suggestion
*Work in progress*
```

File: subtest/docs/documentation.md
Comment (bowni) [inline subtest/docs/documentation.md:46]: We have a risk this might be a bit unclear for a newcomer: maybe explain that adding a dataset means creating a dataset module (like for already existing ones), and that the dataset constructor feeds from it

**Review by bowni — CHANGES_REQUESTED** [review]
Wow, heavy lifting there @arthurPignet (kudos to @aygalic of course) 😎

**Comment by bowni** [comment]
While we are at it: looking at it now it becomes quite clear that we need to refactorize the documentation. I would suggest:

1. A main `README.md` for the description of the project and this package. That is roughly the actual one, **minus** the description of the parameters and the contributivity measurement methods
1. The new `documentation.md` (with more links in it)
1. The tutorials
1. The description of the parameters
1. The description of the contributivity measurement methods

What do you think @arthurPignet @natct10 @RomainGoussault @aygalic ?

**Comment by aygalic** [comment]
I believe some link should be left for another PR once every files (such as notebooks) are present in the master branch. 
Otherwise it will be impossible to test before merging as path will differ.

Current path are : 

`https://github.com/SubstraFoundation/distributed-learning-contributivity/tree/refacto-to-pypi-package-v2/notebooks/examples`

But when merged it will be something like this : 

`https://github.com/SubstraFoundation/distributed-learning-contributivity/notebooks/examples`

I totally agree with the documentation refacto.
I would suggest creating a different PR dedicated to this documentation and reworking the current `README.md`.

File: subtest/scenario.py
Comment (arthurPignet) [reply subtest/scenario.py:98]: Yes it seems possible, as there is no call to self.partner_list in the init. Only run() calls functions which need this attribut
Moved

File: tests/unit_tests.py
Comment (arthurPignet) [reply tests/unit_tests.py:176]: it does not change anything, the full block looks like 
var= (
indented_or_not
)

File: subtest/docs/documentation.md
Comment (arthurPignet) [reply subtest/docs/documentation.md:46]: I do not think that using a new dataset requieres a new module
1- if the package is installed via pip, finding the folder containing the other dataset's modules is not trivial. 
2- A good way to add a new dataset is presented in the example notebooks, in an object oriented way

File: main.py
Comment (RomainGoussault) [inline main.py:19]: what is `subtest` ?

File: setup.py
Comment (RomainGoussault) [inline setup.py:15]: Is `long_description` defined elsewhere ?

File: subtest/docs/documentation.md
Comment (RomainGoussault) [inline subtest/docs/documentation.md:125]: missing space

File: subtest/docs/documentation.md
Comment (RomainGoussault) [inline subtest/docs/documentation.md:118]: these variables names are not super clear, we should either change them or add a quick description

File: subtest/docs/documentation.md
Comment (RomainGoussault) [inline subtest/docs/documentation.md:143]: I would like to see the plot :)

**Review by RomainGoussault — CHANGES_REQUESTED** [review]
Wow huge work @arthurPignet @aygalic !

**Comment by RomainGoussault** [comment]
> While we are at it: looking at it now it becomes quite clear that we need to refactorize the documentation. I would suggest:
> 
> 1. A main `README.md` for the description of the project and this package. That is roughly the actual one, **minus** the description of the parameters and the contributivity measurement methods
> 2. The new `documentation.md` (with more links in it)
> 3. The tutorials
> 4. The description of the parameters
> 5. The description of the contributivity measurement methods
> 
> What do you think @arthurPignet @natct10 @RomainGoussault @aygalic ?

I agree. I would advocate for a small `README.md` (less than a screen) and move most of the stuff in the `documentation.md` page.

File: main.py
Comment (arthurPignet) [reply main.py:19]: it is the name of the package (temporary). At the end, the library would be installed via pip

 pip install subtest

File: setup.py
Comment (arthurPignet) [reply setup.py:15]: Yes, lines 3-4

File: subtest/docs/documentation.md
Comment (arthurPignet) [reply subtest/docs/documentation.md:143]: You mean, in the doc ?

Comment (arthurPignet) [reply subtest/docs/documentation.md:143]: You can run the notebook 3_Exploring_result

Comment (RomainGoussault) [reply subtest/docs/documentation.md:143]: Yes in the doc !

File: setup.py
Comment (RomainGoussault) [reply setup.py:15]: Ah yes my bad

File: subtest/docs/documentation.md
Comment (arthurPignet) [reply subtest/docs/documentation.md:118]: As most of the comments are asking for improvement and modification in the documentation, I think we should merge this one, and open a new one for the documentation

File: subtest/docs/documentation.md
Comment (arthurPignet) [reply subtest/docs/documentation.md:143]: ok, so another request for documentation's improvement

File: subtest/docs/documentation.md
Comment (arthurPignet) [reply subtest/docs/documentation.md:125]: > > While we are at it: looking at it now it becomes quite clear that we need to refactorize the documentation. I would suggest:
> > 
> > 1. A main `README.md` for the description of the project and this package. That is roughly the actual one, **minus** the description of the parameters and the contributivity measurement methods
> > 2. The new `documentation.md` (with more links in it)
> > 3. The tutorials
> > 4. The description of the parameters
> > 5. The description of the contributivity measurement methods
> > 
> > What do you think @arthurPignet @natct10 @RomainGoussault @aygalic ?
> 
> I agree. I would advocate for a small `README.md` (less than a screen) and move most of the stuff in the `documentation.md` page.



> > While we are at it: looking at it now it becomes quite clear that we need to refactorize the documentation. I would suggest:
> > 
> > 1. A main `README.md` for the description of the project and this package. That is roughly the actual one, **minus** the description of the parameters and the contributivity measurement methods
> > 2. The new `documentation.md` (with more links in it)
> > 3. The tutorials
> > 4. The description of the parameters
> > 5. The description of the contributivity measurement methods
> > 
> > What do you think @arthurPignet @natct10 @RomainGoussault @aygalic ?
> 
> I agree. I would advocate for a small `README.md` (less than a screen) and move most of the stuff in the `documentation.md` page.

As most of the comments are asking for improvement and modification in the documentation, I think we should merge this PR, and open a new one for the documentation

**Comment by RomainGoussault** [comment]
Do we have an easy way to make comments on the notebooks ? @bowni @arthurPignet @natct10 
I do have comments but I can't find a good place where to write them.

**Comment by arthurPignet** [comment]
> Do we have an easy way to make comments on the notebooks ? @bowni @arthurPignet @natct10
> I do have comments but I can't find a good place where to write them.

hmm I don't know another way than regular git review

File: subtest/docs/documentation.md
Comment (bowni) [reply subtest/docs/documentation.md]: Yes

**Review by bowni — APPROVED** [review]
To me we are good - the next step is a PR focused on the documentation!

**Review by natct10 — COMMENTED** [review]
I have been able to run:
```python
from subtest.scenario import Scenario
current_scenario = Scenario(partners_count=3,
                            amounts_per_partner=[0.2, 0.5, 0.3],
                            epoch_count=20,
                            minibatch_count=3,
                            dataset_name='mnist')
current_scenario.run()
```

But I'm getting an error with `ruamel` when installing the python package:
```sh
pip3 install -i https://test.pypi.org/simple/ subtest --force-reinstall
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://test.pypi.org/simple/
Collecting subtest
  Downloading https://test-files.pythonhosted.org/packages/5a/77/5fd75c8f6217dd2b528f0d28fd560baff49b4947b4cd2bf57ba37ed1b8f0/subtest-0.0.0.11-py3-none-any.whl (51 kB)
     |████████████████████████████████| 51 kB 166 kB/s 
ERROR: Could not find a version that satisfies the requirement ruamel.yaml==0.16.10 (from subtest) (from versions: none)
ERROR: No matching distribution found for ruamel.yaml==0.16.10 (from subtest)
```

File: subtest/datasets/dataset.py
Comment (bowni) [inline subtest/datasets/dataset.py:78]: Hmm, we can question that. It makes the method less resilient, as it imposes a test on the `dataset_proportion` prior to calling it. Intuitively I'd suggest handling the `== 1` case by returning the entire sets

File: subtest/scenario.py
Comment (bowni) [inline subtest/scenario.py:215]: These hardcoded figures are arbitrary, so I am thinking we should probably put them in `constants`

**Review by bowni — APPROVED** [review]
Still approving although see 2 detail comments:

File: subtest/datasets/dataset.py
Comment (arthurPignet) [reply subtest/datasets/dataset.py:78]: I approuve, I change that

**Comment by arthurPignet** [comment]
> I have been able to run:
> 
> ```python
> from subtest.scenario import Scenario
> current_scenario = Scenario(partners_count=3,
>                             amounts_per_partner=[0.2, 0.5, 0.3],
>                             epoch_count=20,
>                             minibatch_count=3,
>                             dataset_name='mnist')
> current_scenario.run()
> ```
> 
> But I'm getting an error with `ruamel` when installing the python package:
> 
> ```shell
> pip3 install -i https://test.pypi.org/simple/ subtest --force-reinstall
> Defaulting to user installation because normal site-packages is not writeable
> Looking in indexes: https://test.pypi.org/simple/
> Collecting subtest
>   Downloading https://test-files.pythonhosted.org/packages/5a/77/5fd75c8f6217dd2b528f0d28fd560baff49b4947b4cd2bf57ba37ed1b8f0/subtest-0.0.0.11-py3-none-any.whl (51 kB)
>      |████████████████████████████████| 51 kB 166 kB/s 
> ERROR: Could not find a version that satisfies the requirement ruamel.yaml==0.16.10 (from subtest) (from versions: none)
> ERROR: No matching distribution found for ruamel.yaml==0.16.10 (from subtest)
> ```

After numerous tests, I think the error came from the fact that, as we pip install from the test-pypi, pip tries to install all the dependencies from the test-pypi.
 HOWEVER, some of the dependencies listed in the requirements.txt are not avaible in test-pypi, but only on the real pypi

We could confirm this hypothesis when the package will be published on the real pypi. 
By the time, please be sure that you have all the requierements installed via pip before installing subtest

**Comment by bowni** [comment]
Missing 1 approval to be able to merge - @RomainGoussault @natct10 @Thomas-Galtier maybe?

**Review by RomainGoussault — APPROVED** [review]
Great work @arthurPignet !
We should work on the notebooks on the documentation PR.

**Pull request merged**
